compiler/lquantizer/src/LQBinarizer.cpp

```diff
@@ -0,0 +1,192 @@
+      quantization_thresholds[i] = (curr_v + last_v) / 2.0f;
+      last_v = curr_v;
+    }
+  }
+}
+
+// this bin search returns t[l] for values in (t[l], t[l+1]]
+int32_t LQBinarizer::bin_search_encoding(float value)
+{
+  int32_t left = 0, right = quantized_levels - 1;
+
+  while (left < right)
+  {
+    uint32_t middle = (left + right + 1) >> 1;
+
+    if (quantization_thresholds[middle] < value)
+    {
+      left = middle;
+    }
+    else
+    {
+      right = middle - 1;
+    }
+  }
+
+  return left;
+}
+
+void LQBinarizer::quantize_and_pack(const float *data_vector)
+{
+  Shape shape = {encode_bits, data_bin_size};
+  memset(data_binary.get(), 0, shape.num_elements() * sizeof(int32_t));
+
+  for (int32_t i = 0; i < data_float_size; ++i)
+  {
+    // encode value of input
+    int32_t encoding = bin_search_encoding(data_vector[i]);
+
+    // pack value to binary buffer
+    {
+      int32_t idx = i >> 5;    // divide to 32
+      int32_t offset = i & 31; // mod of 32
+      for (int32_t b = 0; b < encode_bits; ++b)
+      {
+        data_binary[calcOffset(shape, b, idx)] |= ((encoding >> b) & 1) << offset;
+      }
+    }
+  }
+}
+
+std::vector<float> LQBinarizer::dequantize()
+{
+  std::vector<float> dequantized_data(data_float_size, 0);
+
+  // calculate data from binary
+  for (int32_t i = 0; i < data_float_size; ++i)
+  {
+    int32_t idx = i >> 5;    // divide to 32
+    int32_t offset = i & 31; // mod of 32
+
+    for (int32_t b = 0; b < encode_bits; ++b)
+    {
+      int32_t bit_value = (data_binary[b * data_bin_size + idx] >> offset) & 1;
+      dequantized_data[i] += (bit_value == 1) ? scales[b] : -scales[b];
+    }
+  }
+
+  return dequantized_data;
+}
+
+// Method returns bits in float representation for data value with index i
+std::vector<float> LQBinarizer::value_bits_as_floats(uint32_t i)
+{
+  std::vector<float> bits(encode_bits);
+
+  int32_t idx = i >> 5;    // divide to 32
+  int32_t offset = i & 31; // mod of 32
+
+  for (int32_t b = 0; b < encode_bits; ++b)
+  {
+    int32_t bit_value = (data_binary[b * data_bin_size + idx] >> offset) & 1;
+    bits[b] = (bit_value == 1) ? 1.0f : -1.0f;
+  }
+
+  return bits;
+}
+
+void LQBinarizer::gradient_descent_scales(const float *data_vector)
+{
+  // iterative gradient descent to search optimal scales `v: argmin||Bxv - x||^2`, `x` - date_vector
+  for (int32_t step = 0; step < descent_steps; ++step)
+  {
+    std::vector<float> Bxv = dequantize();
+    std::vector<float> dv(encode_bits, 0.0f);
+
+    // search dv
+    for (uint32_t i = 0; i < Bxv.size(); ++i)
+    {
+      // define gradient and correct overall step
+      std::vector<float> grad = value_bits_as_floats(i);
+
+      float err = Bxv[i] - data_vector[i];
+      float coef = descent_lr / dequantize().size();
+
+      for (int32_t b = 0; b < encode_bits; ++b)
+      {
+        grad[b] *= err;
+        dv[b] -= coef * grad[b]; // add antigradient with coef to dv
+      }
+    }
```

Comment:
![image](https://user-images.githubusercontent.com/43731397/118337212-f9863a00-b51b-11eb-836e-a0d8924864b3.png)

Power of step by gradient